R/interval_collapse.R

```diff
@@ -0,0 +1,290 @@
+#'   intervals available for each combination of `id_cols` in a dataset.
+#'   Aggregates the collapsed dataset to the common set of intervals.
+#'
+#' @inheritParams agg
+#' @param dt \[`data.table()`\]\cr
+#'   Dataset containing the interval variable.
+#' @param col_stem \[`character(1)`\]\cr
+#'   The name of the variable to collapse, should not include the '_start' or
+#'   '_end' suffix for the interval variable.
+#' @param missing_dt_severity \[`character(1)`\]\cr
+#'   How severe should the consequences of missing intervals that prevent
+#'   collapsing to the most detailed common set of intervals be? Can be either
+#'   'stop', 'warning', 'message', or 'none'. If not "stop", then only the
+#'   intervals that can be correctly collapsed will be done.
+#'
+#' @return \[`data.table()`\] with `id_cols` and `value_cols` columns but with
+#'   the `col_stem` intervals reduced to only the most detailed common set of
+#'   intervals.
+#'
+#' @examples
+#' id_cols <- c("year_start", "year_end", "sex", "age_start", "age_end")
+#' value_cols <- c("value")
+#'
+#' # set up test input data.table
+#' input_dt_male <- data.table::CJ(year_start = 2005, year_end = 2010,
+#'                                 sex = "male",
+#'                                 age_start = seq(0, 95, 5),
+#'                                 value = 25)
+#' input_dt_male[age_start == 95, value := 5]
+#' input_dt_female <- data.table::CJ(year_start = 2005:2009,
+#'                                   sex = "female",
+#'                                   age_start = seq(0, 95, 1),
+#'                                   value = 1)
+#' gen_end(input_dt_female, setdiff(id_cols, c("year_end", "age_end")),
+#'         col_stem = "year", right_most_endpoint = 2010)
+#' input_dt <- rbind(input_dt_male, input_dt_female)
+#' gen_end(input_dt, setdiff(id_cols, "age_end"), col_stem = "age")
+#' data.table::setkeyv(input_dt, id_cols)
+#'
+#'
+#' collapsed_dt <- collapse_common_intervals(
+#'   dt = input_dt,
+#'   id_cols = id_cols,
+#'   value_cols = value_cols,
+#'   col_stem = "year"
+#' )
+#' collapsed_dt <- collapse_common_intervals(
+#'   dt = collapsed_dt,
+#'   id_cols = id_cols,
+#'   value_cols = value_cols,
+#'   col_stem = "age"
+#' )
+#'
+#' @export
+collapse_common_intervals <- function(dt,
+                                      id_cols,
+                                      value_cols,
+                                      col_stem,
+                                      agg_function = sum,
+                                      missing_dt_severity = "stop",
+                                      drop_present_aggs = FALSE) {
+
+  # Validate arguments ------------------------------------------------------
+
+  # check `col_stem` argument
+  assertthat::assert_that(assertthat::is.string(col_stem),
+                          msg = "`col_stem` must be a string")
+  cols <- paste0(col_stem, "_", c("start", "end"))
+
+  # check `agg_function` argument
```

Comment:
How computationally expensive are assertions? Are some of these repeated assertions when this function is run inside `agg` or `scale`?

Reply:
I think most of these are quick, but I did consider adding some option to avoid the assertions.

Will look into this more when doing profiling